core/shared/src/main/scala/scalaz/zio/internal/Env.scala

```diff
@@ -0,0 +1,94 @@
+// Copyright (C) 2018 - 2019 John A. De Goes. All rights reserved.
+package scalaz.zio.internal
+
+import scalaz.zio._
+import scalaz.zio.ExitResult.Cause
+import java.util.concurrent.atomic.AtomicLong
+
+/**
+ * An environment provides the capability to execute different types
+ * of tasks.
+ */
+trait Env {
+
+  /**
+   * Retrieves the executor for the specified type of tasks.
+   */
+  def executor(tpe: Executor.Type): Executor
+
+  /**
+   * Retrieves the scheduler.
+   */
+  def scheduler: Scheduler
+
+  /**
+   * Determines if a throwable is non-fatal or not.
+   */
+  def nonFatal(t: Throwable): Boolean
```

Comment:
with an `unapply`.

Reply:
Nope. `unapply` allocates memory which is undefined in the presence of, e.g. `OutOfMemoryException`.